src/libDirectoryService/Coinbase.cpp

```diff
@@ -175,34 +254,56 @@ void DirectoryService::InitCoinbase() {
 
   LOG_GENERAL(INFO, "Total reward: " << total_reward);
 
+  uint256_t lookupReward = (total_reward / 100) * LOOKUP_REWARD;
+  uint256_t nodeReward = total_reward - lookupReward;
   uint256_t reward_each;
+  uint256_t reward_each_lookup;
 
-  if (!SafeMath<uint256_t>::div(total_reward, sig_count, reward_each)) {
+  if (!SafeMath<uint256_t>::div(nodeReward, sig_count, reward_each)) {
     LOG_GENERAL(WARNING, "reward_each dividing unsafe!");
     return;
   }
 
-  LOG_GENERAL(INFO, "Each reward: " << reward_each);
+  if (!SafeMath<uint256_t>::div(lookupReward, lookup_count,
+                                reward_each_lookup)) {
+    LOG_GENERAL(WARNING, "reward_each_lookup dividing unsafe");
+    return;
+  }
+
+  LOG_GENERAL(INFO, "Each reward: " << reward_each << " lookup each "
+                                    << reward_each_lookup);
 
   uint256_t suc_counter = 0;
+  uint256_t suc_lookup_counter = 0;
   for (auto const& epochNum : m_coinbaseRewardees) {
     LOG_GENERAL(INFO, "[CNBSE] Rewarding " << epochNum.first << " epoch");
 
     for (auto const& shardId : epochNum.second) {
       LOG_GENERAL(INFO, "[CNBSE] Rewarding " << shardId.first << " shard");
-
-      for (auto const& addr : shardId.second) {
-        if (!AccountStore::GetInstance().UpdateCoinbaseTemp(
-                addr, genesisAccount, reward_each)) {
-          LOG_GENERAL(WARNING, "Could Not reward " << addr);
-        } else {
-          suc_counter++;
+      if (shardId.first == -2) {
+        for (auto const& addr : shardId.second) {
+          if (!AccountStore::GetInstance().UpdateCoinbaseTemp(
+                  addr, genesisAccount, reward_each_lookup)) {
+            LOG_GENERAL(WARNING, "Could Not reward " << addr);
+          } else {
+            suc_lookup_counter++;
+          }
+        }
+      } else {
+        for (auto const& addr : shardId.second) {
+          if (!AccountStore::GetInstance().UpdateCoinbaseTemp(
+                  addr, genesisAccount, reward_each)) {
+            LOG_GENERAL(WARNING, "Could Not reward " << addr);
+          } else {
+            suc_counter++;
+          }
         }
       }
     }
   }
 
-  uint256_t balance_left = total_reward - (suc_counter * reward_each);
+  uint256_t balance_left = total_reward - (suc_counter * reward_each) -
```

Comment:
Just checking: Is it possible - especially with the division operations earlier - that somehow `suc_counter * reward_each` and `suc_lookup_counter * reward_each_lookup` end up being a little more than `total_reward`?  In that case, we have an overflow here.  Also, should we use safe math?

Reply:
Already used safemath to divide to get the each reward values